backend/src/main/java/com/openelements/maven/initializer/backend/service/MavenWrapperService.java

```diff
@@ -0,0 +1,179 @@
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package com.openelements.maven.initializer.backend.service;
+
+import com.openelements.maven.initializer.backend.exception.ProjectServiceException;
+import java.io.IOException;
+import java.io.InputStream;
+import java.net.URI;
+import java.nio.file.FileSystem;
+import java.nio.file.FileSystems;
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.nio.file.StandardCopyOption;
+import java.nio.file.attribute.PosixFilePermission;
+import java.util.Collections;
+import java.util.Set;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import org.springframework.stereotype.Service;
+
+/** Service responsible for adding Maven Wrapper (mvnw) files to a generated project. */
+@Service
+public class MavenWrapperService {
+
+  private static final Logger logger = LoggerFactory.getLogger(MavenWrapperService.class);
+
+  private static final String MAVEN_WRAPPER_VERSION = "3.3.4";
+  private static final String MAVEN_VERSION = "3.9.11";
+
+  private static final String MAVEN_WRAPPER_DISTRIBUTION_URL =
+      "https://repo.maven.apache.org/maven2/org/apache/maven/wrapper/maven-wrapper-distribution/"
+          + MAVEN_WRAPPER_VERSION
+          + "/maven-wrapper-distribution-"
+          + MAVEN_WRAPPER_VERSION
+          + "-only-script.zip";
+
+  private static final String MAVEN_DISTRIBUTION_URL =
+      "https://repo.maven.apache.org/maven2/org/apache/maven/apache-maven/"
+          + MAVEN_VERSION
+          + "/apache-maven-"
+          + MAVEN_VERSION
+          + "-bin.zip";
+
+  /**
+   * Adds Maven Wrapper files to the specified project directory.
+   *
+   * <p>This includes:
+   *
+   * <ul>
+   *   <li>Creating the {@code .mvn/wrapper} directory
+   *   <li>Downloading and extracting {@code mvnw} and {@code mvnw.cmd} scripts
+   *   <li>Generating {@code maven-wrapper.properties} with the correct distribution URL
+   * </ul>
+   *
+   * <p>Uses the "only-script" distribution type, which avoids including {@code maven-wrapper.jar}
+   * and downloads Maven directly on first use.
+   *
+   * @param projectRoot the root directory of the generated project
+   * @throws ProjectServiceException if adding the Maven Wrapper fails due to I/O or network issues
+   */
+  public void addMavenWrapper(Path projectRoot) {
+    logger.info("Adding Maven Wrapper to project at: {}", projectRoot);
+
+    try {
+      Path wrapperDir = projectRoot.resolve(".mvn/wrapper");
+      Files.createDirectories(wrapperDir);
+
+      downloadAndExtractWrapper(projectRoot);
+      createWrapperProperties(wrapperDir);
+
+      logger.info("Maven Wrapper added successfully");
+    } catch (IOException e) {
+      logger.error("Failed to add Maven Wrapper", e);
+      throw new ProjectServiceException("Failed to add Maven Wrapper: " + e.getMessage(), e);
```

Comment:
Looking at the tests, I saw that the message is important. For me it is a sign that we should introduce a separate exception.